app/services/create_github_repo_service.rb

```diff
@@ -0,0 +1,208 @@
+      raise Result::Error, error.message
+    end
+
+    owned_private_repos = github_organization_plan[:owned_private_repos]
+    private_repos       = github_organization_plan[:private_repos]
+
+    return true if owned_private_repos < private_repos
+    raise Result::Error, errors(:private_repos_not_available, github_organization_plan)
+  end
+  # rubocop:enable Metrics/AbcSize
+
+  # Public: Add user/team to the GitHubRepository based on the type of assignment.
+  #         Calls #add_user_to_github_repository! if it is an Assignment.
+  #         Calls #add_group_to_github_repository! if it is a GroupAssignment.
+  #
+  # github_repository - GitHubRepository in which we need to add collaborator.
+  #
+  # Returns true if collaborator added or raises a Result::Error.
+  def add_collaborator_to_github_repository!(github_repository)
+    send("add_#{entity.humanize}_to_github_repository!", github_repository)
+  rescue GitHub::Error => error
+    raise Result::Error.new errors(:collaborator_addition_failed), error.message
+  end
+
+  # Maps the type of error to a Datadog error
+  #
+  def report_error(err)
+    case err
+    when /^#{errors(:repository_creation_failed)}/
+      stats_sender.report(:repository_creation_failed)
+    when /^#{errors(:collaborator_addition_failed)}/
+      stats_sender.report(:collaborator_addition_failed)
+    when /^#{errors(:starter_code_import_failed)}/
+      stats_sender.report(:starter_code_import_failed)
+    else
+      stats_sender.report(:failure)
+    end
+  end
+
+  private
+
+  # Internal: Creates a new team on GitHub and adds it to the repository.
+  #
+  # github_repository - GitHubRepository in which we need to add the team.
+  #
+  # Returns true if collaborator added or raises a GitHub::Error.
+  def add_group_to_github_repository!(github_repository)
+    github_team = GitHubTeam.new(organization.github_client, entity.collaborator.github_team_id)
+    github_team.add_team_repository(github_repository.full_name, repository_permissions)
+  end
+
+  # Internal: Creates a new invitation for the GitHubRepository and then accepts it on behalf
+  #           of the user.
+  #
+  # github_repository - GitHubRepository in which we need to add the user.
+  #
+  # Returns true if collaborator added or raises a GitHub::Error.
+  def add_user_to_github_repository!(github_repository)
+    invitation = github_repository.invite(entity.slug, repository_permissions)
+    entity.collaborator.github_user.accept_repository_invitation(invitation.id) if invitation.present?
+  end
+
+  def repository_permissions
+    {}.tap do |options|
+      options[:permission] = "admin" if entity.admin?
+    end
+  end
+
+  # Internal: Method for error messages, modifies error messages based on the type of assignment.
+  #
+  # error_message - A symbol for getting the  appropriate error message.
+  # rubocop:disable LineLength
+  # rubocop:disable MethodLength
+  # rubocop:disable AbcSize
+  def errors(error_message, options = {})
```

Comment:
Can this method be re-organized as a case statement? Then we're not initializing and interpolating a bunch of unneeded strings every time this gets called.